lockstore/lockstore.go

```diff
@@ -326,31 +356,92 @@ func (ls *MemStore) randomHeight() int {
 	return h
 }
 
-func (ls *MemStore) Delete(key []byte) bool {
+func (ls *MemStore) calculateRecomputeHeight(key []byte, hint *Hint, listHeight int) int {
+	recomputeHeight := 0
+	arena := ls.getArena()
+	if hint.height < int32(listHeight) {
+		// Either splice is never used or list height has grown, we recompute all.
+		hint.prev[listHeight] = ls.head
+		hint.next[listHeight] = nil
+		hint.height = int32(listHeight)
+		recomputeHeight = listHeight
+	} else {
+		for recomputeHeight < listHeight {
+			prevNext := hint.prev[recomputeHeight].getNextNode(arena, recomputeHeight)
+			if prevNext != hint.next[recomputeHeight] {
+				recomputeHeight++
+			} else if hint.prev[recomputeHeight] != ls.head &&
+				hint.prev[recomputeHeight] != nil &&
+				bytes.Compare(key, hint.prev[recomputeHeight].getKey(arena)) <= 0 {
+				// Key is before splice.
+				bad := hint.prev[recomputeHeight]
+				for bad == hint.prev[recomputeHeight] {
+					recomputeHeight++
+				}
+			} else if hint.next[recomputeHeight] != nil && bytes.Compare(key, hint.next[recomputeHeight].getKey(arena)) > 0 {
+				// Key is after splice.
+				bad := hint.next[recomputeHeight]
+				for bad == hint.next[recomputeHeight] {
+					recomputeHeight++
+				}
+			} else {
```

Comment:
A bit complex with many `if` conditions in a single statement,  maybe it's easy to miss some conditions which should not go in this `else` path
how about split some condition judges
```
for recomputeHeight < listHeight {
    if  hint not valid {
        recomputeHeight++
    } else {
        if edge condition should stop {
            break
        }
        if Key is before splice {
            change recomputeHeight
        } else if  key is after splice {
            change recomputeHeight
        } else {
            found valid recomputeHeight for this key, break
        }
    }
}
```